File: packages/whisker-firebase/core/ios/Sources/WhiskerFirebaseCore/FirebaseCoreModule.swift
import Foundation
import FirebaseCore
import WhiskerModule

@WhiskerModule
public final class FirebaseCoreModule: Module {
    public override func definition() -> ModuleDefinition {
        ModuleDefinition {
            Name("FirebaseCore")
            Function("initialize") { (_: [WhiskerValue]) -> WhiskerValue in
                if Thread.isMainThread { return Self.initializeApp() }
                return DispatchQueue.main.sync { Self.initializeApp() }
            }
        }
    }

    private static func initializeApp() -> WhiskerValue {
        if FirebaseApp.app() == nil {
            guard let path = Bundle.main.path(forResource: "GoogleService-Info", ofType: "plist"),
                  let options = FirebaseOptions(contentsOfFile: path) else {
                return failure("invalid-config", "Missing or invalid GoogleService-Info.plist in the application bundle")
            }
            FirebaseApp.configure(options: options)
        }
        guard let app = FirebaseApp.app(), let projectID = app.options.projectID else {
            return failure("invalid-config", "Firebase application has no project ID")
        }
        return .map(["value": .map([
            "name": .string(app.name),
            "app_id": .string(app.options.googleAppID),
            "project_id": .string(projectID),
            "storage_bucket": app.options.storageBucket.map { WhiskerValue.string($0) } ?? .null,
        ])])
    }

    private static func failure(_ code: String, _ message: String) -> WhiskerValue {
        .map(["error": .map(["code": .string(code), "message": .string(message)])])
    }
}

File: packages/whisker-firebase/core/android/src/main/kotlin/rs/whisker/modules/firebasecore/FirebaseCoreModule.kt
package rs.whisker.modules.firebasecore

import com.google.firebase.FirebaseApp
import rs.whisker.runtime.Module
import rs.whisker.runtime.ModuleDefinition
import rs.whisker.runtime.WhiskerModule
import rs.whisker.runtime.WhiskerValue

@WhiskerModule
class FirebaseCoreModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("FirebaseCore")
        Function("initialize") {
            try {
                // FirebaseInitProvider initializes from Google Services resources before the Activity.
                val app = FirebaseApp.getInstance()
                val projectId = app.options.projectId
                    ?: throw IllegalStateException("Firebase application has no project ID")
                WhiskerValue.Map(mapOf("value" to WhiskerValue.Map(mapOf(
                    "name" to WhiskerValue.Str(app.name),
                    "app_id" to WhiskerValue.Str(app.options.applicationId),
                    "project_id" to WhiskerValue.Str(projectId),
                    "storage_bucket" to (app.options.storageBucket?.let { WhiskerValue.Str(it) } ?: WhiskerValue.Null),
                ))))
            } catch (error: IllegalStateException) {
                WhiskerValue.Map(mapOf("error" to WhiskerValue.Map(mapOf(
                    "code" to WhiskerValue.Str("invalid-config"),
                    "message" to WhiskerValue.Str(error.message ?: "Firebase initialization failed"),
                ))))
            }
        }
    }
}
